### ISSUE TITLE
Distributed optimization example? 

### ISSUE THREAD

**Issue by feevos** [SheffieldML/GPyOpt#172]
Dear all, 
Is there a distributed optimization example using GPyOpt? 

Thank you for the great work you've put into this.

**Comment by apaleyes** [comment]
Hi! Can you please elaborate a bit more about what use case do you have in mind? Just to make sure everyone understands same thing under "distributed optimization".

**Comment by feevos** [comment]
Hi @apaleyes , thank you very much for your reply. 

I am using GPyOpt for hyperparameter optimization of a deep learning (convNet NN) framework. For example two of these hyperparameters are the learning rate and weight decay of the loss function during training. Currently, I have a "serial" (with respect to the GPyOpt params) solution to the problem: I've created a ```DL_optimizer``` class in which I encapsulate (as members/methods) both the NN as well as the optimization process. For each proposed parameter point of the  ```GPyOpt.methods.BayesianOptimization``` (say x = {x0, x1}) I can use 4 GPUs (in parallel, for increasing batch size) to evolve the network. Therefore, the DL training is in parallel, but the evaluation of the cost function in GPyOpt is serial.  I have access to an HPC cluster (each node has 4 GPUs), and what I hope for is to be able to run, in different nodes, a function evaluation for different values of the proposed hyperparameters. This will divide my computation time for each evaluation as many times as the number of available nodes (approximately). The DL framework I am using is [mxnet](https://mxnet.incubator.apache.org/). I understand that in the Bayesian framework we need to complete the evaluation of the first point, before we propose the (conditioned on the first) second parameter point, but distributed evaluation will allow me to find better solutions in the same time: instead of evaluating a single proposed point, say ```x2 ~P(x2|x1)``` I can have a set of points ```x2_1, x2_2, ... , x2_nodes``` that will more accurately describe the underlying distribution of parameters. 

Now, it is my understanding that distributed evaluation can take place in two ways: a) from within the GPyOpt library (perhaps? I haven't seen an example but **this would be my preferred method because each proposed point would be conditioned to previously evaluated points** - the Bayesian way), or b) by manually distributing the ```DL_optimizer``` class in various nodes, running ```DL_optimizer``` in parallel (so the proposed points of node1, will not have any information of the proposed points of node2), and select the best parameter from both nodes (not ideal). For (b) I am going to try to use the [ray](https://github.com/ray-project/ray) library and I will try and find if possible some way of combining - information from all nodes before proposing new points for all nodes (i.e. before GPyOpt run is complete). I do not know how to do so yet. 

(a) is what I am mostly interested in if such a functionality exists within GPyOpt. Any pointers for the (b) solution (how I can do so manually - in terms of mathematics -, papers? links?) will also be extremely appreciated. 

Thank you very much for your time - by the way I am new in Bayesian optimization so I hope everything I describe makes sense.

**Comment by feevos** [comment]
update: I guess this is one way of doing this (b), by using [external_objective_evaluation](https://github.com/SheffieldML/GPyOpt/blob/master/manual/GPyOpt_external_objective_evaluation.ipynb)

**Comment by apaleyes** [comment]
Well, you pretty much answered the question yourself, external evaluation example is exactly what we have for such use cases. Note that you can use "ignored" points to make sure that you are not revisiting points are are already being evaluated in other nodes.

Admittedly this requries some overhead of controlling the evaluations on different nodes, collecting the results and keeping track of what points are evaluated.

**Comment by feevos** [comment]
Thank you very much @apaleyes , closing this issue.
